api/app/signals/apps/questionnaires/services/question_graph.py

```diff
@@ -0,0 +1,79 @@
+# SPDX-License-Identifier: MPL-2.0
+# Copyright (C) 2021 Gemeente Amsterdam
+import networkx
+
+from signals.apps.questionnaires.app_settings import MAX_QUESTIONS
+from signals.apps.questionnaires.models import Edge, Question
+
+
+class QuestionGraphService:
+    def __init__(self, q_graph):
+        self.q_graph = q_graph
+
+    def load_question_graph_data(self):
+        # Retrieve all relevant edges, questions and answers
+        self.edges = self._get_edges(self.q_graph)
+        self.nx_graph = self._build_nx_graph(self.q_graph, self.edges)
+        self.questions = self._get_all_questions(self.nx_graph)
+
+        # setup caches for quick access
+        self.edges_by_id = {e.id: e for e in self.edges}
+        self.questions_by_id = {q.id: q for q in self.questions}
+
+    def _get_edges(self, q_graph):
+        return list(Edge.objects.filter(graph=q_graph).select_related('choice'))
+
+    @staticmethod
+    def _build_nx_graph(q_graph, edges):
```

Comment:
I tried to add as little information as I could get away with, because this graph will be built for every request involving a session. The reason we need it is to guard against corrupted questionnaires, to have the ability to only show data relevant to the current path through the questionnaire / question graph, etc.

However I think adding these labels should not be a problem, and will add a comment as to why they are there. I will try to find a nice way of making the graph available (this is a similar issue as `can_freeze` below).

Reply:
Info is added now, we will still have to deal with labels that are not printable (possibly by requiring choices to always have a label that is printable).